Api guidelines (#176)
commercetools/commercetools-api-reference: API-GUIDELINE.md

 - Uniformity => checking in the code if there are other properties with similar name and adjust in base of it or improve it
 - Explicability => explain with a few words the purpose of the property avoiding abbreviations (there are some *exceptions).

##Scope
The purpose of these guidelines is to achieve the definition of consistent practices and patterns for all API endpoints.

##URI
Our URIs are composed by:
   
    Base Uri like:
        https://api.{region}.{cloudProvider}.commercetools.com 

    {projectKey}

    resources

    by ID

    by external ID with equals e.g. customers/key={key}

See our documentation for more details https://docs.commercetools.com/http-api.html#hosts

##Resources
Resource is an object or representation of something, which has some associated data with it and can be helped by methods to operate on it.

A resource can be an entity or a collection:
 - Entity is a resource that has to be in singular.
 - Collections are a set of resources that must be in plural.
 
A resource may contain sub-collection resources.

##Methods
The HTTP verbs used are;

 - GET -> retrieve data
 - POST -> create or update data
 - DELETE -> delete data

> It's important to keep in mind the concept of Idempotency. So for instance use GET to retrieve data and not to delete content!

Reviewer: ```suggestion
It's important to avoid unexpected behavior. So for instance use GET to retrieve data and not to delete content!
```

Idempotency is something different and mostly affects only update operations. Most of our updates are not idempotent. Cause either they produce errors or sending twice the same update doesn't result in the same response.

 - use lowercase and separated by hyphens (-) for path segments like /shopping-lists
 - use American English
 - create each resource as unique resource name
 - use the proper HTTP methods for API operations whenever possible
 
 DON'TS
 - DON'T use snake case (_) or other punctuation
 - DON'T use file extensions
 - DON'T use abbreviations
 - DON'T use CRUD HTTP methods or verbs

###Naming Properties
The properties are identified in base of the type: 
 - DateTime
 - Boolean
 - Number/Integer
 - String
 - Array
 - Object

####DateTime
In general, define a DateTime property using a name ending with "At", then if there is a range of date use "From" and "To".

Some examples:

        returnAt
        expirationAt
        validFrom
        validTo

####Boolean
Define a Boolean property using a name which includes "is" or "has" and the verb at past participle.

Some examples:

        isEmailVerified
        hasStagedChanges
        isPublished
        isActive
        
####Number/Integer
Define a Number/Integer property using names that can be quantified.

Some examples:

        availableQuantity
        restockableInDays
        version
        weight
        height

>         orderNumber

Reviewer: orderNumber is a string in our API ;)

        variantId
        totalAmount
      
####String
Define a String property using names that have to be explicable.

Some examples:

      orderEditResult
      productName
      productKey
      productSku
      discountCode
         
####Array

> Define an Array property using names ending with "s".

Reviewer: ```suggestion
Define an Array property using names in plural form, typically ending with "s".
```

There are mass nouns which have no singular form e.g. Money

Some examples:

    discountCodes
    refusedGifts
    roles
    allowedValues
    
####Object

> Define an Object property using "ResourceIdentifier" or "Draft" or "Reference" in the name.

Reviewer: Below examples contradict this statement.

It's more that for objects there are not that much rules.

Draft is used for objects which are meant to be written

ResourceIdentifier for object which refer to other resource and are meant to be written

References for objects which refer to other resources and are only read

> #API GUIDELINES

Reviewer 2: ```suggestion
# API GUIDELINES
```
I think you need a space after all `#`, `##`...

Reviewer 3: I saw that too 🤔 but after testing it worked on markdown intellij

Author: I can confirm: in the Preview nothing change, every paragraph doesn't have the '#'

Reviewer 2: thx for checking.
I'would still prefer with a whitespace. It's easier to read the markdown.

Author: ok I'm gonna update for each of them

##Overview

> As commercetools aims to be 100% API First, developer experience for API solution becomes very important for us.

Reviewer 2: > developer experience for API solution
I don't really understand this sentence. What do you mean by that?

Author: I would rephrase saying: "As commercetools aims to be 100% API First, it's important to have a well designed developer experience"

Reviewer 2: "As commercetools is 100% API first, it's important to have a well designed developer experience"
😉

Author: 🥇

Additionally we aim to achieve more automation and improve the review process of new changes in the APIs.

With this in mind, we have created "API Guidelines" that aims to help developers to design APIs:

 - simple
 - consistent 
 - self-explanatory
 
The Keywords to have a good Guidelines are:
 
 - Consistency => following the Guidelines

>  - Uniformity => checking in the code if there are other properties with similar name and adjust in base of it or improve it

Reviewer 4: It is not clear what "adjust(ing) in base of it or improv(ing) it" means. "In base of" presumably means "on the basis of" or "based on"? What do you adjust? Perhaps adjust(ing) "field naming" based on it? The final phrase "or improve it" rather implies that there are other ways to improve the code, if not, maybe we can omit it. I suggest the following:

```suggestion
 - Uniformity => checking in the code if there are other properties with similar names and adjusting field naming based on it
```

>  - Explicability => explain with a few words the purpose of the property avoiding abbreviations (there are some [exceptions](#exceptions)).

Reviewer 4: ```suggestion
 - Explicability => explaining in a few words the purpose of the property avoiding abbreviations (there are some [exceptions](#exceptions)).
```

Just because other items start with -ing.

## Scope

> The purpose of these guidelines is to achieve the definition of consistent practices and patterns for all API endpoints.

Reviewer 4: ```suggestion
The purpose of these guidelines is to define consistent practices and patterns for all API endpoints.
```

##URI
Our URIs are composed by:
   
    Base Uri like:

>         https://api.{region}.{cloudProvider}.commercetools.com 

Reviewer 2: `api` is one service.
We also have `auth` for example.

Author: make sense I'm gonna specify that the URI sample is for the api service

Author: and for the other services there is the link below, what do you think about?

Reviewer 2: which link? https://docs.commercetools.com/http-api.html#hosts ?
Have you seen https://docs.commercetools.com/http-api-authorization#requesting-an-access-token-using-commercetools-oauth2-server ?

Author: yes and yes ;) but I can mention the link of the authorization if this is what you mean

Reviewer 2: My suggestion:
use `https://{service}.{region}.{cloudProvider}.commercetools.com `
and use the 2 links as examples.

    {projectKey}

    resources

    by ID

    by external ID with equals e.g. customers/key={key}

See our documentation for more details https://docs.commercetools.com/http-api.html#hosts and  https://docs.commercetools.com/http-api-authorization#requesting-an-access-token-using-commercetools-oauth2-server 

## Resources

> Resource is an object or representation of something, which has some associated data with it and can be helped by methods to operate on it.

Reviewer 5: Do we want to be that generic or rather be more specific to our business domain?
```suggestion
A Resource is a data collection object or a data representation of an entity in the e-commerce domain, which is specified by data fields and methods to operate on them.
```

A resource can be an entity or a collection:
 - Entity is a resource that has to be in singular.
 - Collections are a set of resources that must be in plural.
 
A resource may contain sub-collection resources.

## Methods

> The HTTP verbs used are;

Reviewer 6: ```suggestion
The HTTP verbs used are:
```

 - GET -> retrieve data

>  - POST -> create or update data

Reviewer 6: IMO it's worth mentioning that a create should return a `201`, whereas an update should return a `200` HTTP status code.

Author: it's just because there is people that use GET for everything ;)

Author: I can fix it ;)

>  - GET -> retrieve data

Reviewer 5: Apart from GraphQL we always retrieve the full resource, right?
```suggestion
 - GET -> retrieve resource
```

Reviewer 7: It's a good suggestion, we are using resource everywhere it will be very nice if the Guideline itself is consistent and following the convention.

>  - POST -> create data (it should return a 201 HTTP status code) or update data (it should return a 200 HTTP status code)

Reviewer 5: ```suggestion
 - POST -> create resource (it should return a 201 HTTP status code) or update resource (it should return a 200 HTTP status code)
```

>  - DELETE -> delete data

Reviewer 5: ```suggestion
 - DELETE -> delete resource
```

It's important to avoid unexpected behavior. So for instance use GET to retrieve data and not to delete content!

##Properties/Payload
Clients interact with a service by exchanging representations of resources.

> An example, how it's right now the Cart payload:

Reviewer 3: and maybe you can rephrase this to *An example of the Cart payload:* without the how and right now, because it seems like a question

> 

Reviewer: Just realized QueryParameters are missing here too.

Some rules to think about:

- naming
- no change of return type based on parameters
- avoid XOR parameters

> Clients interact with a service by exchanging representations of resources.

Reviewer 5: Not only, also by providing URL parameters.

             {
               "quantity": 2,
               "state": {
                 "typeId": "state",
                 "id": "7c2e2694-aefe-43d7-888e-6a99514caaca"
               }
             }
           ],
           "priceMode": "Platform",
           "lineItemMode": "Standard",
           "totalPrice": {
             "type": "centPrecision",
             "fractionDigits": 2,
             "currencyCode": "EUR",
             "centAmount": 5600
           }
         }
       ],
       "store": {
         "typeId": "store",
         "key": "test-key"
       },
       "cartState": "Active",
       "totalPrice": {
         "type": "centPrecision",
         "fractionDigits": 2,
         "currencyCode": "EUR",
         "centAmount": 5600
       },
       "customLineItems": [],
       "discountCodes": [],
       "inventoryMode": "None",
       "taxMode": "Platform",
       "taxRoundingMode": "HalfEven",
       "taxCalculationMode": "LineItemLevel",
       "refusedGifts": [],
       "origin": "Customer"
     }

## Naming convention

### General Guidelines
To use consistent names these are the rules to follow:

 DO'S
 - use nouns instead of verbs to expose resources

>  - use lowercase and separated by hyphens (-) for path segments like /shopping-lists

Reviewer 5: looks like we use lower camel case for property names composed of more than one word?

Reviewer: In path segments not. For query parameters it's something different

Reviewer 5: sure, I mean, we need an additional bullet point about field names

 - use American English
 - create each resource as unique resource name
 - use the proper HTTP methods for API operations whenever possible
 
 DON'TS
 - DON'T use snake case (_) or other punctuation

>  - DON'T use file extensions

Reviewer 7: possibly we can link this sentense to a set of global file extensions

 - DON'T use abbreviations
 - DON'T use CRUD HTTP methods or verbs

###Naming Properties
The properties are identified in base of the type: 
 - DateTime
 - Boolean
 - Number/Integer
 - String
 - Array
 - Object

####DateTime
In general, define a DateTime property using a name ending with "At", then if there is a range of date use "From" and "To".

Some examples:

        returnAt
        expirationAt
        validFrom
        validTo

####Boolean
Define a Boolean property using a name which includes "is" or "has" and the verb at past participle.

Some examples:

        isEmailVerified
        hasStagedChanges
        isPublished

>         isActive

Reviewer 7: @jenschude you mentioned earlier sometime that we can recommend to write it without is or has as it might be redundant when we generate the SDKs.

e.g. active instead of isActive so that we generate a method named isActive based on the name.

WDYT ?

Reviewer: It depends. On the pure API layer it's more speaking as the type can be read from the name.

For our SDKS its different. E.g. java says that a getter for a boolean bean should start with `is`. So we would have to declare mappings to avoid something like `isIsActive` or `getIsActive` (atm this is how it looks like in PHP)

And defacto the whole API already uses these prefixes and can't be easily changed

Author: so are you suggesting that I should change everything without "is"? Assuming that our guidelines is also for the future

Author: and what about something like "emailIsVerified"?

Reviewer: if so it would be `emailVerified` -> singular rule ;)

And if we don't want to codify the state as is as best as possible, but make decisions about how it should be in the future, it should not only be us to do so ;)

Author: I agree but we should suggest something to the backend team, so I would go with `emailVerified`

Reviewer: No we can't change existing fields. We could only propose how it should be done in the future.

Author: yes, this is what I was saying ;)

> Define a Boolean property using a name which NOT includes "is" but it can include "has" and the verb at past participle.

Reviewer 5: ```suggestion
Define a Boolean property by a the past participle form of a verb. 
Do NOT use the auxiliary verb "is" as prefix here, but "has" can be used in combination with a noun.
```

        
#### Number/Integer
Define a Number/Integer property using names that can be quantified.

Some examples:

        availableQuantity
        restockableInDays
        version
        weight
        height
        variantId
        totalAmount
      
#### String
Define a String property using names that have to be explicable about the content.

Some examples:

      orderEditResult
      productName
      productKey

>       productSku

Reviewer 5: so, in this case the acronym "SKU" will only keep the first letter capitalized?

Author: yes, we are not used to have everything in capital letters as a property

Reviewer 7: please follow further discussion here https://github.com/commercetools/api-design-guidelines/issues/3

      discountCode
         
####Array
Define an Array property using names in plural form, typically ending with "s".

Some examples:

    discountCodes
    refusedGifts
    roles
    allowedValues
    
####Object
Define an Object property using names that have to be explicable about the content and the scope.

Some examples:

    shippingMethod
    shippingRateInput
    product
    shippingDetailsToRemove
    discountedPricePerQuantity
    discountedPrice 

> ####*Exceptions

Author: for my point of view yes, because it's a re-call of line 17 ;)

> ####*Exceptions

Reviewer 3: this is so cool, I didn't know that you can reference a section in markdown too 🙏🏻 thank you. But I think you need to do \[exceptions\]\(#exceptions\) in line 17 to make the anchor work

> ####*Exceptions

Author: thanks for the info I'm gonna update it ASAP :)

The only exception allowed is:

>  - write "id" even if is an abbreviation

Reviewer 5: Let's add some paragraph about method names to this. Create methods are pretty straightforward, I think: "CreateResource"

For update actions we still hold onto this convention?
**set**Field: for optional properties, can be unset if value is empty
**change**Field: for mandatory fields, value cannot be empty

Retrieve methods:
**Query Resources** returns PagedQueryResult containing multiple resources (plural)
**Get Resource by Unique Identifier**, returns only the identified Resource (singular)

Unique Identifier can be UUID, Key, customer email, etc.

Reviewer 7: please follow further communication here https://github.com/commercetools/api-design-guidelines/issues/4

 
The Keywords to have a good Guidelines are:
 
 - Consistency => following the Guidelines
 - Uniformity => checking in the code if there are other properties with similar names and adjusting field naming based on it
 - Explicability => explaining in a few words the purpose of the property avoiding abbreviations (there are some [exceptions](#exceptions)).

## Scope
The purpose of these guidelines is to define consistent practices and patterns for all API endpoints.

## URI
Our URIs are composed by:
   
    Base Uri like:
        https://{service}.{region}.{cloudProvider}.commercetools.com 

    {projectKey}

    resources

    by ID

    by external ID with equals e.g. customers/key={key}

See our documentation for more details https://docs.commercetools.com/http-api.html#hosts and  https://docs.commercetools.com/http-api-authorization#requesting-an-access-token-using-commercetools-oauth2-server 

## Resources
A Resource is a data collection object or a data representation of an entity in the e-commerce domain, which is specified by properties and methods to operate on them.

A resource can be an entity or a collection:
 - Entity is a resource that has to be in singular.

>  - Collections are a set of resources that must be in plural.

Reviewer 4: Please ignore this comment if it is clear to developers (after all I think this is an internal document for technical people).

To me " ... are a set ..." sounds suspicious. This was the starting point.

I am not familiar with the technical things, so I am afraid of making unnecessary noise here, but still I do not understand lines 41-43. The lines 39 and 41 explains a resource in terms of an entity and a collection, where as the lines 42 and 43 explains an entity and collections in terms of resources. (eg. in l. 39, A Resource is ... or a data rep. of an entity; in l. 42, Entity is a resource ... )

Also, does the line 43 mean "A collection is a set of more than one resources"? If so, it contradicts the line 41, which in particular says that *a* resource can be a collection.

I am afraid that I simply missed some intended difference between **R**esource (l. 39) and **r**esource (else where) or I simply do not know the technical definition of the term "be in singular/plural". If this is the case, I am sorry.

Author: no problem Tohru, here the clarification in the line 39 there is the definition easily you can say that there is simple resource (it's just one result) = Entity or a collection resource(it's more then one result) = Collection.

Reviewer 4: Thank you! Now I understand that the line 39 is the definition of a **Resource** in terms of data collection object and an entity. The line 39 itself is now clear. 

With this in mind, the line 41 seems to be either a rephrase of the line 39 or an explanation of the behavior of a single **R**esource. 

Either way, the lines 42 and 43 try to explain the words "entity" and "collection" used in the line 41, but it is done so in terms of Resources, which was defined in terms of entity and collection.
 
I am not sure but I am afraid that this is a circular argument or some word (collection, entity or R/resource) is used in more than one meaning.

So, I think there is a room for improvement here. But since I do not know what you mean to say, I can not unfortunately offer a suggestion. If the passage is clear to others, I am happy to let it go, but if not, maybe we can rephrase a bit?

Author: ok, I will think about it, thanks!

Reviewer 4: Just one final suggestion. If this is about the naming convention, you might want to write so more explicitly and perhaps with examples. 

An entity as a reference must be named in singular (eg. ... ).
A collection as a reference must be named in plural (eg. ... ).

Additionally we aim to achieve more automation and improve the review process of new changes in the APIs.

With this in mind, we have created "API Guidelines" that aims to help developers to design APIs:

 - simple
 - consistent 
 - self-explanatory
 
The Keywords to have a good Guidelines are:
 
 - Consistency => following the Guidelines
 - Uniformity => checking in the code if there are other fields with similar names and adjusting field naming based on it
 - Explicability => explaining in a few words the purpose of the field avoiding abbreviations (there are some [exceptions](#exceptions)).

## Scope
The purpose of these guidelines is to define consistent practices and patterns for all API endpoints.

## URI
Our URIs are composed by:
   
    Base Uri like:
        https://{service}.{region}.{cloudProvider}.commercetools.com 

    {projectKey}

    resources

    URI parameters like:
        /{ID}
        /customer-id={customerId}

See our documentation for more details https://docs.commercetools.com/http-api.html#hosts and  https://docs.commercetools.com/http-api-authorization#requesting-an-access-token-using-commercetools-oauth2-server

> ### Query Parameters

Author: @jenschude could you please have a look on the query parameters paragraph?

> The Query Parameters can be defined as the optional key-value and not unique that appear after the question mark(?) in the URL and they can be combined with other parameters with an ampersand(&).

Reviewer: Should be rephrased as it's hard to understand

Author: could you please check again? ;)

In this case the rules to follow are:
 - follow the same naming convention as the properties

>  - no need to change of return type based on the parameters

Reviewer: ```suggestion
 - don't change the response type based on query parameters
```

Do NOT use the auxiliary verb "is" as prefix here, but "has" can be used in combination with a noun.

Some examples:

        emailVerified
        hasStagedChanges
        published
        active
        
#### Number/Integer
Define a Number/Integer property using names that can be quantified.

Some examples:

        availableQuantity
        restockableInDays
        version
        weight
        height
        variantId
        totalAmount
      
#### String
Define a String property using names that have to be explicable about the content.

Some examples:

      orderEditResult
      productName
      productKey
      productSku
      discountCode
         
#### Array
Define an Array property using names in plural form, typically ending with "s".

Some examples:

    discountCodes
    refusedGifts
    roles
    allowedValues
    
#### Object
Define an Object property using names that have to be explicable about the content and the scope.

Some examples:

    shippingMethod
    shippingRateInput
    product
    shippingDetailsToRemove
    discountedPricePerQuantity
    discountedPrice 

#### Exceptions
The only exception allowed is:
 - write "Id" even if is an abbreviation

>  - write "Sku" even if is an abbreviation

Reviewer 8: ```suggestion
 - write "Sku" even if is an acronym
```
